packages/router/utils.ts

```diff
@@ -1351,12 +1351,26 @@ export class DeferredData {
       this.unlistenAbortSignal();
     }
 
-    if (error) {
+    // use argument length to detect error or value resolution
+    if (arguments.length === 3) {
       Object.defineProperty(promise, "_error", { get: () => error });
       this.emit(false, key);
       return Promise.reject(error);
     }
 
+    // If the promise was resolved with undefined, we'll throw an error as you
+    // should always resolve with a value or null
+    if (data === undefined) {
+      Object.defineProperty(promise, "_error", {
+        get: () =>
+          new Error(
+            `Deferred data for key ${key} resolved to \`undefined\`, you must resolve with a value or \`null\`.`
```

Comment:
```suggestion
            `Deferred data for key "${key}" resolved to \`undefined\`, you must resolve with a value or \`null\`.`
```

This will break the unit test above and will need to add the quotes in there too